opm/simulators/flow/BlackoilModelEbosNldd.hpp

```diff
@@ -208,48 +211,62 @@ class BlackoilModelEbosNldd {
         const auto domain_order = this->getSubdomainOrder();
 
         // -----------   Solve each domain separately   -----------
+        DeferredLogger logger;
         std::vector<SimulatorReportSingle> domain_reports(domains_.size());
         for (const int domain_index : domain_order) {
             const auto& domain = domains_[domain_index];
             SimulatorReportSingle local_report;
-            switch (model_.param().local_solve_approach_) {
-            case DomainSolveApproach::Jacobi:
-                solveDomainJacobi(solution, locally_solved, local_report,
-                                  iteration, timer, domain);
-                break;
-            default:
-            case DomainSolveApproach::GaussSeidel:
-                solveDomainGaussSeidel(solution, locally_solved, local_report,
-                                       iteration, timer, domain);
-                break;
+            try {
+                switch (model_.param().local_solve_approach_) {
+                case DomainSolveApproach::Jacobi:
+                    solveDomainJacobi(solution, locally_solved, local_report, logger,
+                                      iteration, timer, domain);
+                    break;
+                default:
+                case DomainSolveApproach::GaussSeidel:
+                    solveDomainGaussSeidel(solution, locally_solved, local_report, logger,
+                                           iteration, timer, domain);
+                    break;
+                }
+            }
+            catch (...) {
+                // Something went wrong during local solves.
+                local_report.converged = false;
             }
             // This should have updated the global matrix to be
             // dR_i/du_j evaluated at new local solutions for
             // i == j, at old solution for i != j.
             if (!local_report.converged) {
                 // TODO: more proper treatment, including in parallel.
-                OpmLog::debug("Convergence failure in domain " + std::to_string(domain.index));
+                logger.debug(fmt::format("Convergence failure in domain {} on rank {}." , domain.index, rank_));
             }
             domain_reports[domain.index] = local_report;
         }
 
-        // Log summary of local solve convergence to DBG file.
+        // Communicate and log all messages.
+        auto global_logger = gatherDeferredLogger(logger, model_.ebosSimulator().vanguard().grid().comm());
+        global_logger.logMessages();
+
+        // Accumulate local solve data.
+        int num_converged = 0;
+        int num_converged_already = 0;
+        int num_local_newtons = 0;
+        int num_domains = domain_reports.size();
```

Comment:
If you put these in an array then you can sum them all across the ranks using a single call of the form
```
comm.sum(convergence_counts, convergence_counts.size());
```
instead of having to issue communication calls for each quantity.  The downside to this, of course, is that it becomes a little harder to read, although you could use
```C++
int& num_converged = convergence_count[0];
// ...
```
and so on to give more meaningful names to each element.